File: lib/csharp/Protocol.cs
using System;
using System.Collections.Generic;
using System.Text;
public class Protocol : IProtocol 
{
	//private Encoding encode = Encoding.GetEncoding("gb2312");
	//private Encoding encode=UTF8Encoding.Default;
    private Encoding encode = Encoding.UTF8;

	public byte[] recvBuffer;
	public byte[] sendBuffer;

	public int recvPos = 0;
	public int sendPos = 0;

	public virtual void  writeMsgBegin(){}
	public virtual void  writeMsgEnd(){}
	/**  Writing voids. */

	private void write(byte[] bytes, int len)
	{
		Array.Copy(bytes, 0, sendBuffer, sendPos, len);
		sendPos+=len;
	}
	public void  writeBool(bool value)
	{
		byte[] bytes = BitConverter.GetBytes(value);
		write(bytes,1);
	}
	
	public void  writeUInt8( byte i)
	{
		sendBuffer[sendPos] = i;
		sendPos+=1;
	}
	public void  writeInt8( sbyte i)
	{
		sendBuffer[sendPos] = (byte)i;
		sendPos+=1;
	}
	
	public void  writeUInt16(ushort i)
	{
		byte[] bytes = BitConverter.GetBytes(i);
		write(bytes,2);
	}
	public void writeInt16(short i)
	{
		byte[] bytes = BitConverter.GetBytes(i);
		write(bytes,2);
	}
	
	public void  writeUInt32( uint i)
	{
		byte[] bytes = BitConverter.GetBytes(i);
		write(bytes,4);
	}
	public void  writeInt32( int i) 
	{
		byte[] bytes = BitConverter.GetBytes(i);
		write(bytes,4);
	}
	
	public void  writeInt64(long i)
	{
		byte[] bytes = BitConverter.GetBytes(i);
		write(bytes,8);
	}
	
	public void  writeFloat(float f)
	{
		byte[] bytes = BitConverter.GetBytes(f);
		write(bytes,4);
	}

	public void writeString(string str)
	{
		byte[] byteData = encode.GetBytes(str);
		byte[] byteLen = BitConverter.GetBytes((short)(byteData.Length));
		write(byteLen,2);
		write(byteData,byteData.Length);
	}
	
	/** Reading voids */
	public bool  readBool()
	{
		bool temp = BitConverter.ToBoolean(recvBuffer, recvPos);
		recvPos+=1;
		return temp;
	}
	
	public byte  readUInt8()
	{
		byte temp = recvBuffer[recvPos];
		recvPos+=1;
		return temp;
	}
	public sbyte  readInt8()
	{
		sbyte temp = (sbyte)(recvBuffer[recvPos]);
		recvPos+=1;
		return temp;
	}
	
	public ushort readUInt16()
	{
		ushort temp = BitConverter.ToUInt16(recvBuffer, recvPos);
		recvPos+=2;
		return temp;
	}
	public short readInt16()
	{
		short temp = BitConverter.ToInt16(recvBuffer, recvPos);
		recvPos+=2;
		return temp;
	}
	
	public uint  readUInt32()
	{
		uint temp = BitConverter.ToUInt32(recvBuffer, recvPos);
		recvPos+=4;
		return temp;
	}
	public int  readInt32()
	{
		int temp = BitConverter.ToInt32(recvBuffer, recvPos);
		recvPos+=4;
		return temp;
	}
	
	public long  readInt64()
	{
		long temp = BitConverter.ToInt64(recvBuffer, recvPos);
		recvPos+=8;
		return temp;
	}
	
	public float  readFloat()
	{
		float temp = BitConverter.ToSingle(recvBuffer, recvPos);
		recvPos+=4;
		return temp;
	}

	public string readString()
	{
		short strlen = readInt16();
        string temp = encode.GetString(recvBuffer, recvPos, strlen);
		recvPos+=strlen;
		return temp;
	}
}

File: lib/go/Protocol.go
//==============================================
/**
@date:		2015:8:26
@file: 		Protocol.go
@author		zhangshuai
*/
//==============================================
package rpc

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"sync"
)

type Protocol struct {
	sync.Mutex //可能会在不同一个协程中
	Rpc        *RpcSession
	RecvBuf    *bytes.Buffer
	SendBuf    *bytes.Buffer
}

func (this *Protocol) WriteMsgBegin() {
	this.Lock()
	this.SendBuf.Reset()
}
func (this *Protocol) WriteMsgEnd() {

	buf := new(bytes.Buffer)
	binary.Write(buf, binary.LittleEndian, uint16(this.SendBuf.Len()))
	buf.Write(this.SendBuf.Bytes())
	this.Rpc.Send(buf)
	this.Unlock()
}

/**  Writing functions. */

func (this *Protocol) WriteBool(b bool) {
	if b {
		this.WriteUInt8(1)
	}
	this.WriteUInt8(0)
}

func (this *Protocol) WriteUInt8(i uint8) {

	binary.Write(this.SendBuf, binary.LittleEndian, i)
}

func (this *Protocol) WriteInt8(i int8) {
	binary.Write(this.SendBuf, binary.LittleEndian, i)

}

func (this *Protocol) WriteUInt16(i uint16) {

	binary.Write(this.SendBuf, binary.LittleEndian, i)

}

func (this *Protocol) WriteInt16(i int16) {

	binary.Write(this.SendBuf, binary.LittleEndian, i)

}

func (this *Protocol) WriteUInt32(i uint32) {

	binary.Write(this.SendBuf, binary.LittleEndian, i)
}

func (this *Protocol) WriteInt32(i int32) {

	binary.Write(this.SendBuf, binary.LittleEndian, i)
}

func (this *Protocol) WriteInt64(i int64) {
	binary.Write(this.SendBuf, binary.LittleEndian, i)

}

func (this *Protocol) WriteFloat(f float32) {

	binary.Write(this.SendBuf, binary.LittleEndian, f)
}

func (this *Protocol) WriteString(str string) {
	this.WriteUInt16(uint16(len(str)))
	var tmp = []uint8(str)
	err := binary.Write(this.SendBuf, binary.LittleEndian, tmp)
	if err != nil {
		fmt.Println(err)
	}
}

/** Reading functions */
func (this *Protocol) ReadBool() bool {
	b := this.ReadUInt8()
	if b == 1 {
		return true
	} else {
		return false
	}
}

func (this *Protocol) ReadUInt8() uint8 {
	var i uint8
	binary.Read(this.RecvBuf, binary.LittleEndian, &i)
	return i
}

func (this *Protocol) ReadInt8() int8 {
	var i int8
	binary.Read(this.RecvBuf, binary.LittleEndian, &i)
	return i
}

func (this *Protocol) ReadUInt16() uint16 {
	var i uint16
	binary.Read(this.RecvBuf, binary.LittleEndian, &i)
	return i
}

func (this *Protocol) ReadInt16() int16 {
	var i int16
	binary.Read(this.RecvBuf, binary.LittleEndian, &i)
	return i
}

func (this *Protocol) ReadUInt32() uint32 {
	var i uint32
	binary.Read(this.RecvBuf, binary.LittleEndian, &i)
	return i
}

func (this *Protocol) ReadInt32() int32 {
	var i int32
	binary.Read(this.RecvBuf, binary.LittleEndian, &i)
	return i
}

func (this *Protocol) ReadInt64() int64 {
	var i int64
	binary.Read(this.RecvBuf, binary.LittleEndian, &i)
	return i
}

func (this *Protocol) ReadFloat() float32 {
	var f float32
	binary.Read(this.RecvBuf, binary.LittleEndian, &f)
	return f
}

func (this *Protocol) ReadString() string {
	n := this.ReadUInt16()
	buf := make([]uint8, n)
	err := binary.Read(this.RecvBuf, binary.LittleEndian, buf)
	if err != nil {
		fmt.Println(err)
	}
	return string(buf)
}
